### PR TITLE
Issue #1727: Improve revert reason when assuming interface compliance contracts.  

### PR DESCRIPTION
**Pull request by NoopurShinde**
* Solved issue #1727: Improve revert reason when assuming interface compliance contracts.

* Fixes

   * Added the below code in ERC721.sol

           bytes memory payload = abi.encodeWithSignature(
            "onERC721Received(address,address,uint256,bytes)",
            msg.sender,
            from,
            tokenId,
            _data
            );
            
            (bool success, bytes memory returndata) = to.call(payload);

            if (!success) {
                if(returndata.length > 0){
                    revert(string (returndata));
                }
                else
                    revert("ERC721: to address does not implement ERC721Received interface");
            } else {
                    return true;
                   }

   * Added 2 dummy contracts in  **/contracts/mocks**; one which does not implement ERC721Received and one which implements it but reverts as the contract rejects the transaction. 

    * Updated the test cases in ERC721.behavior.js to test the code added in ERC721.sol

### CODE REVIEW

**Comment by NoopurShinde** [comment]
Hi @nventuro @frangio 

Any update on the PR?

**Comment by NoopurShinde** [comment]
Hi @nventuro @frangio 

Any update on the PR?

File: package.json [hunk 1]
```diff
@@ -53,9 +53,9 @@
     "eslint-config-standard": "^11.0.0",
     "eslint-plugin-import": "^2.18.2",
     "eslint-plugin-mocha-no-only": "^1.1.0",
-    "eslint-plugin-node": "^5.2.1",
+    "eslint-plugin-node": "^10.0.0",
```
Comment (frangio) [inline package.json]: Can you please undo the changes in `package.json`? We prefer to have PRs that focus on separate issues.

File: contracts/token/ERC721/ERC721.sol [hunk 2]
```diff
@@ -331,8 +331,24 @@ contract ERC721 is Context, ERC165, IERC721 {
             return true;
         }
 
-        bytes4 retval = IERC721Receiver(to).onERC721Received(_msgSender(), from, tokenId, _data);
-        return (retval == _ERC721_RECEIVED);
+        bytes memory payload = abi.encodeWithSignature(
+            "onERC721Received(address,address,uint256,bytes)",
+            msg.sender,
```
Comment (frangio) [inline contracts/token/ERC721/ERC721.sol]: Please use `_msgSender()` instead, for compatibility with `Context`. This is used for integration with the [GSN](https://gasstation.network).

```suggestion
            _msgSender(),
```

File: contracts/token/ERC721/ERC721.sol [hunk 3]
```diff
@@ -331,8 +331,24 @@ contract ERC721 is Context, ERC165, IERC721 {
             return true;
         }
 
-        bytes4 retval = IERC721Receiver(to).onERC721Received(_msgSender(), from, tokenId, _data);
-        return (retval == _ERC721_RECEIVED);
+        bytes memory payload = abi.encodeWithSignature(
+            "onERC721Received(address,address,uint256,bytes)",
```
Comment (frangio) [inline contracts/token/ERC721/ERC721.sol]: Please use `abi.encodeWithSelector` with `IERC721Receiver(to).onERC721Received.selector` as the first argument. This ensures that we don't mess up the string constant.

File: contracts/mocks/ERC721ReceiverNotImplementedMock.sol [hunk 4]
```diff
@@ -0,0 +1,14 @@
+pragma solidity ^0.5.0;
+
+
+contract ERC721ReceiverNotImplementedMock{
+    bytes4 private _retval;
+    bool private _reverts;
```
Comment (frangio) [inline contracts/mocks/ERC721ReceiverNotImplementedMock.sol]: Please remove these two variables since they're not used.

File: contracts/mocks/ERC721ReceiverRevertsMock.sol [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+    constructor (bytes4 retval, bool reverts) public {
+        _retval = retval;
+        _reverts = reverts;
+    }
+
+    function onERC721Received(address operator, address from, uint256 tokenId, bytes memory data)
+        public returns (bytes4)
+    {
+        require(!_reverts, "ERC721ReceiverMock: Transaction rejected by receiver");
```
Comment (frangio) [inline contracts/mocks/ERC721ReceiverRevertsMock.sol]:
```suggestion
        require(!_reverts, "ERC721ReceiverRevertsMock: Transaction rejected by receiver");
```

File: contracts/token/ERC721/ERC721.sol [hunk 6]
```diff
@@ -331,8 +331,24 @@ contract ERC721 is Context, ERC165, IERC721 {
...
+            from,
+            tokenId,
+            _data
+            );
+
+            //solhint-disable-next-line avoid-low-level-calls
+        (bool success, bytes memory returndata) = to.call(payload);
+        if (!success) {
+            if (returndata.length > 0){
+                revert(string(returndata));
```
Comment (frangio) [inline contracts/token/ERC721/ERC721.sol]: From reading the Solidity documentation, it seems that both `revert` and `require` abi-encode their arguments as if they were a call to an `Error(string)` function.

If I understand correctly, this means that we're "wrapping" the revert reason twice, since `returndata` is already encoded here. I think we have to drop to assembly and use the `revert` opcode directly to avoid this re-encoding and to bubble up the value as-is.

Let me know if this was clear @NoopurShinde.

File: contracts/token/ERC721/ERC721.sol [hunk 7]
```diff
@@ -331,8 +331,24 @@ contract ERC721 is Context, ERC165, IERC721 {
...
+            }else 
+                revert("ERC721: to address does not implement ERC721Received interface");
+        } else {return true;} 
+        // bytes4 retval = IERC721Receiver(to).onERC721Received(_msgSender(), from, tokenId, _data);
+        // return (retval == _ERC721_RECEIVED);
```
Comment (frangio) [inline contracts/token/ERC721/ERC721.sol]: Please tidy up the whitespace here to conform to our usual style and remove the commented out lines.  :slightly_smiling_face:
